dae/dae/genomic_resources/__init__.py

```diff
@@ -1,24 +1,88 @@
+from typing import Dict, Callable
+from importlib_metadata import entry_points, EntryPoint
+
 from .repository_factory import build_genomic_resource_repository
 from .repository import GenomicResource
+from .resource_implementation import GenomicResourceImplementation
+
+
+_FOUND_RESOURCE_IMPLEMENTATIONS: Dict[str, EntryPoint] = {}
+
+
+_REGISTERED_RESOURCE_IMPLEMENTATIONS: Dict[
+    str, Callable[[GenomicResource], GenomicResourceImplementation]] = {}
 
-# from .embeded_repository import GenomicResourceEmbededRepo
-# from .url_repository import GenomicResourceURLRepo
-# from .dir_repository import GenomicResourceDirRepo
 
 __all__ = [
     "build_genomic_resource_repository", "GenomicResource",
+    "get_resource_implementation_factory"
 ]
 
 
-# register_real_genomic_resource_repository_type(
-#     "url", GenomicResourceURLRepo)
-# register_real_genomic_resource_repository_type(
-#     "directory", GenomicResourceDirRepo)
-# register_real_genomic_resource_repository_type(
-#     "embeded", GenomicResourceEmbededRepo)
+_IMPLEMENTATIONS_LOADED = False
+_PLUGINS_LOADED = False
 
 
-_PLUGINS_LOADED = False
+def get_resource_implementation_factory(
+    resource_type: str
+) -> Callable[[GenomicResource], GenomicResourceImplementation]:
+    """
+    Return an implementation builder for a certain resource type.
+
+    If the builder is not registered, then it will search for an entry point
+    in the found implementations list. If an entry point is found, it will be
+    loaded and registered and returned.
+    """
+    if resource_type not in _REGISTERED_RESOURCE_IMPLEMENTATIONS:
+        if resource_type not in _FOUND_RESOURCE_IMPLEMENTATIONS:
+            raise ValueError(
+                f"unsupported resource implementation type: {resource_type}"
+            )
+        entry_point = _FOUND_RESOURCE_IMPLEMENTATIONS[resource_type]
+        loaded = entry_point.load()
+        register_implementation(resource_type, loaded)
+        _REGISTERED_RESOURCE_IMPLEMENTATIONS[resource_type] = loaded
```

Comment:
I may be misreading this, but isn't this line a repeat of what is done in `register_implementation`?

Reply:
whoops

Reply:
Fixed